src/Microsoft.AspNetCore.SignalR/DefaultHubLifetimeManager.cs

```diff
@@ -106,6 +106,9 @@ public override Task InvokeUserAsync(string userId, string methodName, object[]
 
         public override Task OnConnectedAsync(HubConnectionContext connection)
         {
+            // Set the hub groups feature
+            connection.Features.Set<IHubGroupsFeature>(new HubGroupsFeature());
```

Comment:
Pit of failure for who exactly? This runs before application code runs and it's private to the implementation of this implementation of the HubLifetimeManager.